### PR TITLE
do/ci: Improve github `workflow_call`

### PR DESCRIPTION
**Pull request by phlax**
- Print diskspace at beginning/end of jobs
- Add diskspace hack for jobs that need it
- Add pre/post steps
- Add `env` parser for reusable workflows
- Separate `do_ci` to an action so it can be reused more easily

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: .github/workflows/_ci.yml [hunk 1]
```diff
@@ -52,15 +73,47 @@ jobs:
...
+        df -h
+
+        TO_REMOVE=(
+            /opt/hostedtoolcache
+            /usr/local/lib/android
+            /usr/local/.ghcup)
+
+        for removal in "${TO_REMOVE[@]}"; do
+            echo "Removing: ${removal} ..."
+            # sudo rm -rf "$removal"
```
Comment (phlax) [inline .github/workflows/_ci.yml:96]:
```suggestion
            sudo rm -rf "$removal"
```

File: .github/workflows/_ci.yml [hunk 2]
```diff
@@ -93,3 +146,16 @@ jobs:
       env:
         GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
         ENVOY_DOCKER_BUILD_DIR: ${{ runner.temp }}
+
+    - run: |
+        echo "disk space at end of build:"
+        df -h
+        echo
+        du -ch "$(Build.StagingDirectory)" | grep -E "[0-9]{2,}M|[0-9]G"
```
Comment (phlax) [inline .github/workflows/_ci.yml:154]:
```suggestion
        du -ch "${{ github.tmp }}" | grep -E "[0-9]{2,}M|[0-9]G"
```

File: .github/workflows/_ci.yml [hunk 3]
```diff
@@ -52,15 +73,47 @@ jobs:
...
+      uses: envoyproxy/toolshed/gh-actions/using/recurse@actions-v0.0.7
+      with:
+        uses: ${{ inputs.run_pre }}
+        with: ${{ inputs.run_pre_with }}
+
     - id: do_ci
       name: 'Run CI target ${{ inputs.target }}'
       run: |
+        if [[ -n "${{ inputs.env }}" ]]; then
+            echo "${{ inputs.env }}" > /tmp/sourcevars
```
Comment (phlax) [inline .github/workflows/_ci.yml:113]: use a tempfile

**Comment by phlax** [comment]
tested here: https://github.com/envoyproxy/envoy/actions/runs/5581348944/jobs/10199401909

File: .github/workflows/_ci.yml [hunk 4]
```diff
@@ -52,44 +73,64 @@ jobs:
...
-        ENVOY_DOCKER_BUILD_DIR: ${{ runner.temp }}
+
+    - run: |
+        echo "disk space at end of build:"
+        df -h
+        echo
+        du -ch "${{ runner.temp }}" | grep -E "[0-9]{2,}M|[0-9]G"
+      name: "Check disk space at end"
+
+    - if: ${{ inputs.run_post }}
```
Comment (phlax) [inline .github/workflows/_ci.yml:132]: perhaps this should go before the diskspace dump above

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks. 

Although the test workflow was failed, but it should have nothing to do with the ci script. 🤔

**Pull request merged**